### ISSUE TITLE
How to handle power failures?  Not really an issue, just want to pick your brain.

### ISSUE THREAD

**Issue by fayena** [BlakeCLewis/PILN#2]
I've been thinking of how to handle a power failure.       

I've got the pilnfired.py set to run as a service and I've added code to check for finished segments.   If the segment is finished, it steps to the next segment.   It should work for the up swing, as if your past you're your low temp holds then you should be fine to go ahead and fire.   If you've got a hold on the way down and power is cut while holding, it should be ok to redo the hold.   

Am I missing anything?      

Might be easier to just handle the power flickers with a battery backup on the raspberry pi.

**Comment by BlakeCLewis** [comment]
I have killed the daemon, changed some variables and then restarted the daemon. I think it picked up on the current segment and continued. If outage is short, things may work out.

Preferred new feature:
Text my phone alerting me that there are issues, so I can make decisions.
When I have had issues, I fix the issue, copy the current profile, edit it, cutting out stuff that has been completed, then start the new profile.  A text message alert could wake draw my attention to deal with a problem. I can set up the phone to alert me every 30 sec until I read message, which is how my oncall work phone is setup.

Past issues:
-  Trial of PID algorithm changes, lots of these;
-   Loose thermocouple connections, heat cycle expansion/contraction can loosen up compression screws, indicated by downward temp spikes that are happening faster than kiln normally cools. I now solder thermocouple wires to the breakout board, the cheap terminals are crap. On the thermocouple side, I put thermocouple and wire under same screw, direct wire to thermocouple contact;
-   Shorted out an element wire to grounded case, which stuck/welded relay in ON position, blew breaker, trashed relay;
-   I chose a glaze profile for a bisque fire, oops, stopped run, edited bisque profile to pick up and continued with new profile.

All of the kiln manufactures state "Do Not Fire Unattended!". I stay with my kiln. I stay with it until power to kiln elements is disabled. I might leave to use the bathroom or eat dinner, but I bring up the kiln controller on my phone and watch. I usually fire at night. I sleep in the kiln shed. I make sure I have fresh air to breath. I set an alarm to wake me every 30 minutes.

There are failure modes, when not monitored, that could be catastrophic e.g. stuck relays. Some of the failure modes could be monitored, but If you follow the requirement "Do Not Fire Unattended!" then they are taken care of by the human. Even the commercial computer control kilns state  "Do Not Fire Unattended!". I like my using controller, I made it, I can change it, but I know that it is not better at failure modes than commercial kiln controllers whom state "Do Not Fire Unattended!".

**Comment by pvarney** [comment]
Yes! Text on issues. Have just recently fired the kiln back up to do 160 plates for our wedding and have had some issues with a bad connection on my primary relay board. I was thinking that adding some texting would be really good. I'd be interested in that if you code it. 

Interesting observation on the thermocouple connection. I had some problems with inconsistent temperatures a while back and am wondering if that was the issue. I re-did the connections and haven't had any issues this time around.

I may be stealing this code back since I'm interesting in using the kiln-sitter as a sensor. My elements are old and cost more to replace than what I paid for the kiln. So it has a tough time getting up to cone 6 temperature and sat at a lower temp for too long, over firing the glazes we were testing. Bought some cones so that I can see what the temperature cure looks like when it hits a real cone 6.

Thanks for sharing your experiences.

**Comment by fayena** [comment]
I added some code that allows the daemon to check to see which segments are completed and start the segment that is not completed.    Seems to work great for quick power flickers.    

Before I implemented this, the daemon would start over so if it was in segment 2 at 800C it would drop the temp back down to the segment 1 temp then start to ramp backup.    I also added a bit of code that would flag the firing as an error if you get a +200C degree difference between the read temp and the ramp temp.     So, if your kiln elements are wore out and it's simply not reaching temp, it will flag and exit rather than keep trying to reach temperature.   

I finally got my website up and running so I'll be back to working on this and doing several more firings shortly.

**Comment by BlakeCLewis** [comment]
I been reading up on thermocouple wires and solder.
Solder doesn't stick to the thermocouple wires. It does a sort of mechanical junction "encapsulation". I might have to go back to screw terminals. I re-soldered the wires today to get good reading. There is a company(Tinkerforge #2109) that makes MAX31856 module with thermocouple socket, but it is expensive.   Mouser sells them.

**Comment by fayena** [comment]
I was getting temps consistently 20-30C low.   Well 4 firings so not a huge sample but consistently in the same range.   I "think" it was a cold junction error caused by the wiring junctions on the thermocouple being a different material than the type k thermocouple wires (aka cheap thermocouple).       I ended up putting my thermocouple extension wires directly in the hole with the thermocouple wires instead of using the junctions on the thermocouple.   I just stuck them under the thermocouple wires and then used the screwed to secure.  That really seemed to help.     I need to do more firings to make sure but I'm hopeful since I was within 5C last firing.

**Comment by BlakeCLewis** [comment]
I had a power failure last glaze fire. The kilnsitter triggered and it was a few minutes into a 15min hold. It was programmed to drop 56c and hold for 30min.  Everything looked good. Everythin on the hog is goot. The linemen showed up 20 minutes after outage started and put a new fuse in the line. I think it had dropped over 200C(360F) in that 20 minutes. I was lucky it wasn't 30 sooner.
